### ISSUE TITLE
[ocaml5-issue] Segfault on s390x in Effect test

### ISSUE THREAD

**Issue by jmid** [ocaml-multicore/multicoretests#383]
After merging #381 to `main` I spotted a multicoretests-ci failure on s390x for `5.2+trunk`
https://ocaml-multicoretests.ci.dev:8100/job/2023-08-17/073031-ci-ocluster-build-e8d54d#L1589
```
random seed: 97230569
generated error  fail  pass / total     time test name

[ ]     0     0     0     0 / 20000     0.0s Lin DSL ref int test with Effect
[ ]     0     0     0     0 / 20000     0.0s Lin DSL ref int test with Effect (generating)
[✓] 20000     0     0 20000 / 20000     0.4s Lin DSL ref int test with Effect

[ ]     0     0     0     0 / 20000     0.0s Lin DSL ref int64 test with Effect
[✓] 20000     0     0 20000 / 20000     0.4s Lin DSL ref int64 test with Effect

[ ]     0     0     0     0 / 20000     0.0s Lin DSL CList int test with Effect
[✓] 20000     0     0 20000 / 20000     0.4s Lin DSL CList int test with Effect

[ ]     0     0     0     0 / 20000     0.0s Lin DSL CList int64 test with Effect
[✓] 20000     0     0 20000 / 20000     0.5s Lin DSL CList int64 test with Effect

[ ]     0     0     0     0 / 20000     0.0s negative Lin DSL ref int test with EffectFile "src/neg_tests/dune", line 92, characters 7-27:
92 |  (name lin_tests_dsl_effect)
            ^^^^^^^^^^^^^^^^^^^^
(cd _build/default/src/neg_tests && ./lin_tests_dsl_effect.exe --verbose)
Command got signal SEGV.
```

After focusing and cleaning up I now have a reproducer in the https://github.com/ocaml-multicore/multicoretests/commits/s390x-crash-repro branch that consistently fails in the CI on both `5.1+trunk` and `5.2+trunk`.

I will open an OCaml issue for it shortly.

**Comment by jmid** [comment]
Reported upstream in https://github.com/ocaml/ocaml/issues/12486

**Comment by jmid** [comment]
Closing as this was fixed and cherry-picked to `5.1.0~rc3` in https://github.com/ocaml/ocaml/pull/12530
